Automatically add a nonce to script tag if a csp-nonce meta tag is available to get it from

TylerRick opened the pull request:
Resolves #100

TylerRick committed:
Add www.example.com to config.hosts

This resolves a bunch of (6) test failures that were failing due to an error page that said:

  To allow requests to www.example.com, add the following to your environment configuration:
  config.hosts << "www.example.com"

This was added as part of ActionDispatch::HostAuthorization in Rails 6. See:
- https://guides.rubyonrails.org/6_0_release_notes.html#action-pack-notable-changes
- https://guides.rubyonrails.org/configuring.html#configuring-middleware

TylerRick committed:
Automatically add a nonce to script tag (#100)

... if a csp-nonce meta tag is available to get it from.

- Added test that confirms that nonce is added
- Added an end-to-end test that uses a strict CSP. Confirmed that this fails without the nonce added
  to javascript_include_tag calls.
- With Sprockets 4, it doesn't add a separate script tag for jquery, so add an explicit
  javascript_include_tag "jquery" for that case

Co-authored-by: Matt Brictson <mattbrictson@users.noreply.github.com>

mattbrictson commented on spec/dummy/app/controllers/application_controller.rb:
```diff
@@ -1,9 +1,17 @@
   def root
   end
 
+  def strict_csp
```
Where this controller action is being used?

TylerRick replied:
Ah, good catch! I needed this action to duplicate the actual problem (JS errors), but I guess I hadn't added a test that used it yet.

Since `spec/xray/middleware_spec.rb` (where it was) only checks implementation details (that a certain `nonce` attribute is added) but doesn't actually test that this is enough to prevent JS/CSP errors, I added a quick end-to-end (Webdriver) test that verifies this in a real browser. Seems to work...

mattbrictson commented:
Thanks for the PR! I left some initial comments. I won't be able to approve or merge this PR until we get Travis CI working.

@brentd or @TylerRick do you know what we aren't seeing CI checks on this PR?

mattbrictson commented:
I manually restarted the Travis CI build and it seems to be working now 🤔

mattbrictson commented:
@TylerRick This is failing CI because CI is testing against Rails < 6. I think the appropriate next steps are:

1. Open a separate PR to add Rails 6 to the CI matrix
2. Update this PR to only run CSP-related test code for Rails >= 6.0.0

Can you help with either or both of those?

TylerRick commented on lib/xray/middleware.rb:
```diff
@@ -110,10 +110,15 @@ def script_matcher(script_name)
     end
 
+    def nonce_from_meta_tag(html)
+      (match = html.match(/<meta name="csp-nonce" content="([^"]*)"/)) && match[1]
```
Nice! I'll have to remember that one. So much more concise.

TylerRick committed:
Test on Rails 6

Spread out the Rails versions in travis.yml a bit and test 5.2, not just 5.0

TylerRick committed:
Add end-to-end test that actually tests in a real browser that pressing Ctrl+Shift+X makes the Xray
UI appear

This verifies that there are no JavaScript errors that would prevent it from working, even though
the xray.js script was injected into the document (as tested by spec/xray/middleware_spec.rb).

TylerRick committed:
Update script_matcher to allow for .source and .debug suffixes that sometimes get added by Sprockets
4 pipelines

TylerRick commented on spec/dummy/app/controllers/application_controller.rb:
```diff
@@ -1,9 +1,17 @@
 
+  content_security_policy only: :strict_csp do |policy|
+    policy.script_src :self, :strict_dynamic
+  end
```
Looks like Rails 5.2.

Yes, good idea.

TylerRick commented on spec/xray/middleware_spec.rb:
```diff
@@ -95,6 +95,12 @@ def mock_response(status, content_type, body)
     expect(page).to have_selector('script[src^="/assets/xray"]')
   end
 
+  it "adds nonce to the script tag" do
```
Thanks!

TylerRick committed:
Locked rspec-rails to 3.x in the .gemspec

(Since Gemfile.lock isn't currently under version control, and probably can't be since RAILS_VERSION
is dynamic.)

Without this, it will install the latest version, 4.0, which uses &. not available in older Rubies, causing errors like this:

   ruby/2.1.0/gems/rspec-rails-4.0.0/lib/rspec/rails/example/mailer_example_group.rb:25: syntax error, unexpected '.'
               options&.each { |key, value| default_u...

TylerRick committed:
Fix test error with Rails 6:

  Sprockets::Railtie::ManifestNeededError:
  Expected to find a manifest file in `app/assets/config/manifest.js` but did not

TylerRick committed:
Add capybara-screenshot so we can check out the .html document when a test fails

TylerRick committed:
Load gems from Gemfile.local too, if present, so that developers can use some optional debugging
tools like byebug and pry

TylerRick committed:
Fix test failures on Rails < 6 due to sprockets >= 4

TylerRick commented:
> @TylerRick This is failing CI because CI is testing against Rails < 6. I think the appropriate next steps are:
> 
> 1. Open a separate PR to add Rails 6 to the CI matrix
> 2. Update this PR to only run CSP-related test code for Rails >= 6.0.0

Good catch. I've updated it to only run CSP-related test code for Rails >= 5.2.

I also added Rails 6 (and 5.2) to the testing matrix. It's a really trivial change. I don't think we need a separate PR for that unless you're likely to not merge this one right away... Also had to fix another issue related to sprockets 4... Everything is in pretty self-contained commits if we need to cherry pick anything.

Now just trying to fix the test failures... hopefully they'll pass this time around :crossed_fingers: 

```
  3) Xray Bar includes the controller and action
     Failure/Error: find('#xray-bar').should have_text('ApplicationController#root')
     
     Capybara::ElementNotFound:
       Unable to find css "#xray-bar"
     # ./spec/xray/xray_bar_spec.rb:7:in `block (2 levels) in <top (required)>'
```

TylerRick committed:
In CI, we can't open the .html files saved by capybara-screenshot, so puts page.html instead so we
can figure out why it failed

TylerRick committed:
Don't test so many old versions

Ruby 2.4 is end-of-life
(https://www.ruby-lang.org/en/news/2020/04/05/support-of-ruby-2-4-has-ended/)

TylerRick commented:
Okay, finally got tests passing. Merge please?

Wasted too much time getting it to work with old Rubies/Rails, when I think the correct path should be to just remove support for them going forward (#109)...

rocket-turtle commented:
Hi @mattbrictson,

I love this gem and wie just upgraded to sprockets 4. https://github.com/brentd/xray-rails/issues/113
If this MR had been merged it would have saved us a lot of time and trouble.

Do you plan to continue merging MR and releaseing new versions for this Gem?
I don't think there is any similar gem like this, so it would be sad to see it fade away.

If I can help in any way please let me know.

rocket-turtle commented:
Hi @brentd,

I just saw that you released a new version for the gem.
Do you plan to release more versions?
Are you willing to accept pull requests?

I would like to contribute to this gem.

Currentle I'm working on my own [branch](https://github.com/rocket-turtle/xray-rails/commits/master/) which is based on this Pull request. 

If you merge this in I can add some value:

* cc3f449 - drop support for Ruby 1.9 - remove xray/aliasing (1 year ago) <Jonas S>
* bf96eea - use rspec-rails 6.1 (1 year ago) <Jonas S>
* e0b837d - use and test rails 6.1 (1 year ago) <Jonas S>
* f5966e1 - fix spec for macs (1 year ago) <Jonas S>
* 0d0ed27 - Convert specs to RSpec 3.9.3 syntax with Transpec (1 year ago) <Jonas S>
* e293c7c - FIX ArgumentError: wrong number of arguments (given 4, expected 3) (1 year ago) <Jonas S>
* e536b07 - remove non working capybara-screenshot (1 year ago) <Jonas S>

Hope to help
Jonas

brentd commented:
Hi @rocket-turtle - since I'm rewriting xray, in the meantime I will accept PRs with small non-breaking changes. From what I can tell your changes seem OK; I can probably make an exception for removing Ruby 1.9 support since your change toward Module prepend is good. 

However, I may end up removing any/ all contributions in the rewrite, so please be aware. Once released I will tag the rewrite version 1.0 and as it will likely not support old versions of Ruby/Rails, pre-1.0 users may still benefit from your changes.
